microsoft/testsuites/kselftest/kselftest.py

```diff
@@ -0,0 +1,160 @@
+from lisa.tools.tar import Tar
+from lisa.util import LisaException, find_patterns_in_lines
+from lisa import notifier, Environment
+from assertpy import assert_that
+
+
+@dataclass
+class KselftestResult:
+    name: str = ""
+    status: TestStatus = TestStatus.QUEUED
+    exit_value: int = 0
+
+
+class Kselftest(Tool):
+    _RESULT_KSELFTEST_OK_REGEX = re.compile(r"ok (.*)\s+")
+    @property
+    def command(self) -> str:
+        return str(self._command)
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        # command to run kselftests
+        self._command = (
+            self.get_tool_path(use_global=True)
+            / "kselftest-packages/run_kselftest.sh -c breakpoints"
+        )
+        # tar file path specified in yml
+        self._tar_file_path = kwargs.pop("kselftest_file_path", "")
+
+        if not self._tar_file_path:
+            raise LisaException(f"tar file not found in: {self._tar_file_path}")
+
+    def _install(self) -> bool:
+        # tool_path corresponds to path /home/lisatest/lisa_working/tool/kselftest
+        tool_path = self.get_tool_path(use_global=True)
+        # file_name corresponds to
+        # /home/lisatest/lisa_working/tool/kselftest/kselftests.tar.xz
+        file_name = tool_path / os.path.basename(self._tar_file_path)
+        # copies tar file from local machine to remote machine
+        self.node.shell.copy(self._tar_file_path, file_name)
+
+        self.node.tools[Tar].extract(file_name, tool_path)
+        self._log.debug(f"INSTALL from {self._tar_file_path}!")
+
+        return self._check_exists()
+
+    def run_all(self, environment: Environment, log_path: str) -> bool:
+        self.run(" > kselftest-results.txt", force_run=True, sudo=True, shell=True)
+        # copy kselftest-results.txt from remote to local node for processing results
+        remote_copy = self.node.tools[RemoteCopy]
+        remote_copy.copy_to_local(
+            PurePosixPath("/home/lisatest/kselftest-results.txt"),
+            PurePath(log_path),
+        )
+
+        local_kselftest_results_path = PurePath(log_path) / "kselftest-results.txt"
+
+        # parse results from local_kselftest_results_path file
+        # read the file
+        with open(local_kselftest_results_path, "r") as f:
+            result_output = f.read()
+            results = self._parse_results(result_output)
+
+         # assert that all tests passed
+        failed_tests = []
+        for result in results:
+            if result.status == TestStatus.FAILED:
+                failed_tests.append(result.name)
+
+            # create test result message
+            info: Dict[str, Any] = {}
+            info["information"] = {}
+            info["information"]["name"] = result.name
+            info["information"]["exit_value"] = result.exit_value
+            subtest_message = create_test_result_message(
```

Comment:
This is incorrect. It should be something like this:
```
subtest_message = create_test_result_message(
                SubTestMessage,
                test_result,
                environment,
                result.name,
                result.status,
                other_fields=info,
            )
```

where test_result is of type `TestResult` and passed from the test suite itself. A complete example is given here: https://github.com/microsoft/lisa/blob/d50773629ee2f36deb15204b9136c3798e3e65d5/microsoft/testsuites/ltp/ltpsuite.py#L86-L93